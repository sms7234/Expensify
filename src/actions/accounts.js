import database from '../firebase/firebase';

export const setAccounts = (accounts) => ({
  type: "SET_ACCOUNTS",
  accounts
});

export const startSetAccounts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts`)
      .once('value')
      .then((snapshot) => {
        const accounts = [];
        snapshot.forEach((child) => {
          accounts.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setAccounts(accounts));
      });
  };
};

const addAccount = (account) =>({
  type: "ADD_ACCOUNT",
  account
});

export const startAddAccount = (accountData={}) => {
  return (dispatch, getState) => {
    const {
      account='',
      description=''
    } = accountData;
    const accountItem={account,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/accounts`).push(accountItem).then((ref)=>{
      dispatch(addAccount({
        id: ref.key,
        ...accountItem
      }));
    });
  };
};

export const editAccount = (id, updates) => ({
  type: 'EDIT_ACCOUNT',
  id,
  updates
});

export const startEditAccount = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts/${id}`).update(updates).then(()=>{
      dispatch(editAccount(id,updates));
    });
  };
};

export const removeAccount = (id) => ({
  type: 'REMOVE_ACCOUNT',
  id
});

export const startRemoveAccount = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts/${id}`).remove().then(()=>{
      dispatch(removeAccount(id));
    });
  };
};
