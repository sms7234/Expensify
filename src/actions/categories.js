import database from '../firebase/firebase';

export const setCategories = (categories) => ({
  type:"SET_CATEGORIES",
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
      .once('value')
      .then((snapshot) => {
        const categories = [];
        snapshot.forEach((child) => {
          categories.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setCategories(categories));
      });
  };
};

export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) =>{
    const {
      category= '',
      description= ''
    } = categoryData;
    const categoryItem = {category, description};
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).push(categoryItem).then((ref)=>{
      dispatch(addCategory({
        id: ref.key,
        ...categoryItem
      }));
    });
  };
};

export const editCategory = (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
});

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(()=>{
      dispatch(editCategory(id,updates));
    });
  };
};

export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
});

export const startRemoveCategory = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=>{
      dispatch(removeCategory(id));
    });
  };
};
