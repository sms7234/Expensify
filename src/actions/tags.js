import database from '../firebase/firebase';

export const setTags = (tags) => ({
  type: "SET_TAGS",
  tags
});

export const startSetTags = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags`)
      .once('value')
      .then((snapshot) => {
        const tags = [];
        snapshot.forEach((child) => {
          tags.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setTags(tags));
      });
  };
};

export const addTag = (tag) =>({
  type: "ADD_TAG",
  tag
});

export const startAddTag = (tagData={}) => {
  return (dispatch, getState) => {
    const {
      tag='',
      description=''
    } = tagData;
    const tagItem={tag,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/tags`).push(tagItem).then((ref)=>{
      dispatch(addTag({
        id: ref.key,
        ...tagItem
      }));
    });
  };
};

export const editTag = (id, updates) => ({
  type: 'EDIT_TAG',
  id,
  updates
});

export const startEditTag = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).update(updates).then(()=>{
      dispatch(editTag(id,updates));
    });
  };
};

export const removeTag = (id) => ({
  type: 'REMOVE_TAG',
  id
});

export const startRemoveTag = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).remove().then(()=>{
      dispatch(removeTag(id));
    });
  };
};
