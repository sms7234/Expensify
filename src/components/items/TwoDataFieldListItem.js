import React from 'react';
import {Link} from 'react-router-dom';

const CategoryListItem=({dispatch, id, value, description, location}) =>(
  <Link className="list-item" to={`/${location}/${id}`}>
    <div>
      <h3 className="list-item__title">{value}</h3>
    </div>
    <h3 className="list-item__data">{description}</h3>
  </Link>
)

export default CategoryListItem;
