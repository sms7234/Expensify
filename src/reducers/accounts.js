export default (state=[], action) => {
  switch(action.type){
    case 'SET_ACCOUNTS':
      return action.accounts;
    case 'ADD_ACCOUNT':
      return state.concat(action.account);
    case 'EDIT_ACCOUNT':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_ACCOUNT':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
