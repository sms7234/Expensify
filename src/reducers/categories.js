export default (state = [], action) => {
  switch(action.type){
    case 'ADD_CATEGORY':
      return state.concat(action.category);
    case 'SET_CATEGORIES':
      return action.categories;
    default:
      return state;
  }
};
