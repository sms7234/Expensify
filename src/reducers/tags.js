export default (state=[], action) => {
  switch(action.type){
    case 'SET_TAGS':
      return action.tags;
    case 'ADD_TAG':
      return state.concat(action.tag);
    case 'EDIT_TAG':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_TAG':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
