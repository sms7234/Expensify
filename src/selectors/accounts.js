export default (accounts, {account}) => {
  return accounts.filter((item) => {
    const textMatch = item.account.toLowerCase().includes(account.toLowerCase());
    return textMatch;
  }).sort((a,b)=> {
    if(a.account < b.account){
      return -1;
    } else{
      return 1;
    }
  });
};
