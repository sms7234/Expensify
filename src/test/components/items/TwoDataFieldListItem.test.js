import React from 'react';
import {shallow} from 'enzyme';
import TwoDataFieldListItem from '../../../components/items/TwoDataFieldListItem';
import accounts from '../../fixtures/accounts';
import categories from '../../fixtures/categories';
import tags from '../../fixtures/tags';

test('should render TwoDataFieldListItem with one item from accounts', () => {
  const wrapper = shallow(<TwoDataFieldListItem key={accounts[0].id}
  id={accounts[0].id}
  value={accounts[0].account}
  description={accounts[0].description}
  location={'editAccount'}
  />)
  expect(wrapper).toMatchSnapshot();
})

test('should render TwoDataFieldListItem with one item from category', () => {
  const wrapper = shallow(<TwoDataFieldListItem key={categories[0].id}
  id={categories[0].id}
  value={categories[0].category}
  description={categories[0].description}
  location={'editCategory'}
  />)
  expect(wrapper).toMatchSnapshot();
})

test('should render TwoDataFieldListItem with one item from tags', () => {
  const wrapper = shallow(<TwoDataFieldListItem key={tags[0].id}
  id={tags[0].id}
  value={tags[0].tag}
  description={tags[0].description}
  location={'editTag'}
  />)
  expect(wrapper).toMatchSnapshot();
})
