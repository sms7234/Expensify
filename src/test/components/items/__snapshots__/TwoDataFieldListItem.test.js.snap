// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TwoDataFieldListItem with one item from accounts 1`] = `
<Link
  className="list-item"
  to="/editAccount/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      BoA
    </h3>
  </div>
  <h3
    className="list-item__data"
  >
    dog costs
  </h3>
</Link>
`;

exports[`should render TwoDataFieldListItem with one item from category 1`] = `
<Link
  className="list-item"
  to="/editCategory/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Dog
    </h3>
  </div>
  <h3
    className="list-item__data"
  >
    dog costs
  </h3>
</Link>
`;

exports[`should render TwoDataFieldListItem with one item from tags 1`] = `
<Link
  className="list-item"
  to="/editTag/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      tax-deductible
    </h3>
  </div>
  <h3
    className="list-item__data"
  >
    tax docs
  </h3>
</Link>
`;
